src/algorithms/calorimetry/TrackClusterMergeSplitter.cc

```diff
@@ -0,0 +1,460 @@
+        debug(
+          "{} clusters to merge: current sum = {}, current significance = {}, {} track(s) pointing to merged cluster",
+          m_mapClustToMerge[clustAndProject.first].size(),
+          eClustSum,
+          sigSum,
+          m_mapProjToMerge[clustAndProject.first].size()
+        );
+      }  // end cluster loop
+    }  // end matched cluster-projection loop
+
+    // do cluster merging
+    for (auto clustToMerge : m_mapClustToMerge) {
+
+      // add clusters to be merged
+      m_vecClust.clear();
+      for (const int& iClustToMerge : clustToMerge.second) {
+        m_vecClust.push_back( (*in_protoclusters)[iClustToMerge] );
+      }
+      m_vecClust.push_back( (*in_protoclusters)[clustToMerge.first] );
+
+      // for each track pointing to merged cluster, merge clusters
+      for (const int& iMatchedTrk : m_mapProjToMerge[clustToMerge.first]) {
+        edm4eic::MutableProtoCluster merged_cluster = out_protoclusters->create();
+        merge_clusters(
+          m_vecProject[iMatchedTrk],
+          m_vecClust,
+          merged_cluster
+        );
+      }
+    }  // end clusters to merge loop
+
+    // copy unused clusters to output
+    for (auto in_cluster : *in_protoclusters) {
+
+      // ignore used clusters
+      if (m_mapIsConsumed[in_cluster.getObjectID().index]) {
+        continue;
+      }
+
+      // copy cluster and add to output collection
+      edm4eic::MutableProtoCluster out_cluster = out_protoclusters->create();
+      copy_cluster(in_cluster, out_cluster);
+
+    }  // end cluster loop
+    trace("Copied unused clusters into output collection");
+
+  }  // end 'process(Input&, Output&)'
+
+
+
+  // --------------------------------------------------------------------------
+  //! Reset bookkeeping containers
+  // --------------------------------------------------------------------------
+  void TrackClusterMergeSplitter::reset_bookkeepers() const {
+
+    m_mapIsConsumed.clear();
+    m_mapClustProject.clear();
+    m_mapClustToMerge.clear();
+    m_mapProjToMerge.clear();
+    m_vecProject.clear();
+    m_vecClust.clear();
+    trace("Reset bookkeeping containers");
+
+  }  // end 'reset_bookkeepers()'
+
+
+
+  // --------------------------------------------------------------------------
+  //! Collect projections pointing to calorimeter
+  // --------------------------------------------------------------------------
+  void TrackClusterMergeSplitter::get_projections(
+    const edm4eic::TrackSegmentCollection* projections,
+    const edm4eic::CalorimeterHit& hit
+  ) const {
+
+    // return if projections are empty
+    if (projections -> size() == 0) {
+      debug("No projections in input collection.");
+      return;
+    }
+
+    // get readout
+    auto readout = m_converter -> findReadout(
+      m_converter -> findDetElement(
+        m_converter -> position(hit.getCellID())
+      )
+    );
+
+    // grab detector id
+    const int id = m_detector -> constant<int>(
```

Comment:
Please, move lookups to the `init`. Also, why not just make a decoder for the `"system"` field and extract it from cellID that way?

Reply:
Like I mentioned up above, an alternative would be to provide the ID name as a parameter. Then we wouldn't even need the decoder...